python/mlc_chat/model/qwen2/qwen2_model.py

```diff
@@ -0,0 +1,276 @@
+                "Overriding %s from %d to %d (%s)",
+                bold("prefill_chunk_size"),
+                self.prefill_chunk_size,
+                self.context_window_size,
+                bold("context_window_size"),
+            )
+            self.prefill_chunk_size = self.context_window_size
+        assert self.tensor_parallel_shards == 1, "QWEN currently does not support sharding."
+
+
+# pylint: disable=invalid-name,missing-docstring,too-many-locals
+
+
+class QWen2Attention(nn.Module):  # pylint: disable=too-many-instance-attributes
+    def __init__(self, config: QWen2Config):
+        head_dim = config.hidden_size // config.num_attention_heads
+
+        self.c_attn = nn.Linear(
+            in_features=config.hidden_size,
+            out_features=(2 * config.num_key_value_heads + config.num_attention_heads) * head_dim,
+            bias=True,
+        )
+        self.o_proj = nn.Linear(
+            config.num_attention_heads * head_dim, config.hidden_size, bias=False
+        )
+        # KV cache for single sequence
+        self.k_cache = nn.KVCache(
+            config.context_window_size, [config.num_key_value_heads, head_dim]
+        )
+        self.v_cache = nn.KVCache(
+            config.context_window_size, [config.num_attention_heads, head_dim]
+        )
+
+        self.hidden_size = config.hidden_size
+        self.head_dim = head_dim
+        self.num_attention_heads = config.num_attention_heads
+        self.num_key_value_heads = config.num_key_value_heads
+        self.rope_theta = config.rope_theta
+
+    def forward(self, hidden_states: Tensor, attention_mask: Tensor, total_seq_len: tir.Var):
+        bsz, sl, _ = hidden_states.shape
+        assert bsz == 1, "Only support batch size 1 at this moment."
+        # Step 1. QKV Projection
+        qkv = self.c_attn(hidden_states)
+        num_heads = 2 * self.num_key_value_heads + self.num_attention_heads
+        qkv = op.reshape(qkv, (bsz, sl, num_heads, self.head_dim))
+        # Step 2. Apply QK rotary embedding
+        q, k, v = op_ext.llama_rope(
+            qkv, total_seq_len, self.rope_theta, self.num_attention_heads, self.num_key_value_heads
+        )
+        # Step 3. Query and update KVCache
+        self.k_cache.append(op.squeeze(k, axis=0))
+        self.v_cache.append(op.squeeze(v, axis=0))
+        k = self.k_cache.view(total_seq_len)
+        v = self.v_cache.view(total_seq_len)
+        # Step 4. Compute softmax(Q @ K^T / sqrt(d)) @ V
+        output = op_ext.attention(q, k, v, casual_mask=attention_mask)
+        # Step 5. Apply output projection
+        return self.o_proj(output)
+
+
+ACT2FN = {
+    "gelu": partial(nn.gelu, approximate=False),
+    "relu": nn.relu,
+    "silu": nn.silu,
+    "swish": nn.silu,
+    "gelu_new": partial(nn.gelu, approximate=True),
+}
+
+
+class QWen2MLP(nn.Module):
+    def __init__(self, config: QWen2Config):
+        self.gate_proj = nn.Linear(config.hidden_size, config.intermediate_size, bias=False)
```

Comment:
Here, we can combine the gate_proj and up_proj matmul similar to what we did in combing QKV in attention layer.
```
self.gate_up_proj = nn.Linear(
    in_features=config.hidden_size,
    out_features=2 * config.intermediate_size,
    bias=False,
)
self.down_proj = nn.Linear(self.intermediate_size, config.hidden_size, bias=False)
```

And in `forward`,
```
concat_x1_x2 = self.gate_up_proj(x)
x1, x2 = op.split(concat_x1_x2, 2, axis=-1)
return self.down_proj(self.act_fn(x1) * x2)
```

Also need to transform the parameters in the loader, similar to this one: https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/model/llama/llama_loader.py#L65